VoodooRMI/Functions/F30.cpp

```diff
@@ -7,95 +7,73 @@
  */
 
 #include "F30.hpp"
-#include <Configuration.hpp>
 
-OSDefineMetaClassAndStructors(F30, RMIFunction)
-#define super RMIFunction
+OSDefineMetaClassAndStructors(F30, RMIGPIOFunction)
+#define super RMIGPIOFunction
 
-bool F30::attach(IOService *provider)
-{
-    int retval = rmi_f30_initialize();
-    
-    if (retval < 0)
+bool F30::init(OSDictionary *dictionary) {
+    if (!super::init(dictionary))
         return false;
-    
-    super::attach(provider);
-    
+
+    query_regs = reinterpret_cast<uint8_t *>(IOMalloc(RMI_F30_QUERY_SIZE * sizeof(uint8_t)));
```

Comment:
This feels like something that should be done for all the functions or none of them. The line of what F30 should do vs the RMIGPIOFunction is a bit confusing.

I think I'll still merge it in though, if you can get the other things fixed.

Reply:
I want to replace the default size to a dynamic one but haven't come up with a good idea.

Reply:
How about the last commit?

Reply:
F30 is pretty complicated. It's got a lot of ctrl registers for haptics and LEDs that I eventually want to mess around with. I think trying to do any sort of inheritance for the registers is going to end up more complicated in the end. I think an interface instead where F30/F3A loops through the each GPIO pin, figures out if it's valid, and then registers it with RMIGPIOFunction might work better? That way the query/ctrl registers stay in F30/F3A and we don't need to alloc anything. Still need to allocate the button array, but we might be able to use a set method in RMIGPIOFunction right before we start registering buttons? If you've got another solution, I'd be happy to hear it. I don't think this is quite it though.